### PR TITLE
pn-validator improvements #pn 2470

### PR DESCRIPTION
**Pull request by AndreaCimini90**
## Short description
- validate an incoming object if it has keys with different types than those defined in the model
- refine tests that test exceptions

## List of changes proposed in this pull request
- Added isString, isObject ... to check the type of the property
- Added not rule to negate rules
- Handled exceptions in tests

## How to test
Create a validator and test the validation in case of the type is different and in case of negated rule

### CODE REVIEW

File: packages/pn-validator/src/rules/ForEachElement.ts [hunk 1]
```diff
@@ -19,38 +26,75 @@ export class ForEachElement<TModel, TValue> extends Rule<TModel, TValue> {
...
+          if (results[index] === undefined) {
+            results.push(resultToAdd);
             continue;
           }
-          results.push(null);
+          results[index] = resultToAdd;
+          index++;
         }
-        if (results.length > 0) {
+        if (results.filter(r => r === null).length > results.length) {
```
Comment (clombardi) [inline packages/pn-validator/src/rules/ForEachElement.ts:97]: I cannot imagine a scenario for this condition to be true, in general I guess that 
```
someArray.filter(someCondition).length <= someArray.length
```
holds for any array and for any condition. But maybe there is some case I am not aware of.

File: packages/pn-validator/src/types/TypeRules.ts [hunk 2]
```diff
@@ -1,43 +1,20 @@
...
-  ? DateRuleValidator<TModel, TValue>
-  : {};
-
-type BoolenaTypeRules<TModel, TValue> = [TValue] extends [Boolean | undefined | null]
-  ? CommonRuleValidator<TModel, TValue>
-  : {};
-
-type ArrayTypeRules<TModel, TValue> = [TValue] extends [Array<infer _TEachValue> | undefined | null]
-  ? ArrayRuleValidator<TModel, TValue>
+export type TypeRules<TModel, TValue> = [TValue] extends [String | undefined | null]
```
Comment (clombardi) [inline packages/pn-validator/src/types/TypeRules.ts:8]: I remember that we tried several variants of this definition. Maybe a comment about what it should be taken with care about any possible changes (i.e. keep the `| undefined | null` in each type, and just one `{}` at the end) would help other people understand the form of this definition.

**Review by clombardi — CHANGES_REQUESTED** [review]
In general, I agree with the proposed changes, and in my opinion the add of the negated variants improves indeed the usability of this nice tool.   
I have some minor comments, I think that the one regarding the maybe-always-true-condition could deserve some analysis.

File: packages/pn-validator/src/ruleValidators/NumberRuleValidator.ts [hunk 3]
```diff
@@ -43,14 +52,30 @@ export class NumberRuleValidator<TModel, TValue>
...
+        customErrorMessage
+      )
+    );
     return this;
   };
+
+  /**
+   * Negate next rule
+   */
+  public readonly not = (): NotRuleValidator<TModel, TValue> =>
```
[the remark this answers is not in the document]
Comment (AndreaCimini90) [reply packages/pn-validator/src/ruleValidators/NumberRuleValidator.ts:77]: This is can be redundant. For condition >= <= there is the Between rule

File: packages/pn-personafisica-webapp/src/validators/PersonValidator.ts [hunk 4]
```diff
@@ -8,8 +8,9 @@ import { Person } from '../redux/delegation/types';
 export class PersonValidator extends Validator<Person> {
     constructor() {
       super();
-      this.ruleFor('firstName').isEqual('Mario');
-      this.ruleFor('lastName').isEmpty(true);
+      this.ruleFor('firstName').isString().isEqual('Mario');
+      this.ruleFor('lastName').isString().isEmpty();
```
[the remark this answers is not in the document]
Comment (AndreaCimini90) [reply packages/pn-personafisica-webapp/src/validators/PersonValidator.ts:12]: Yes, you are right!

File: packages/pn-validator/src/ValidatorBuilder.ts [hunk 5]
```diff
@@ -35,11 +42,42 @@ export class ValidatorBuilder<TModel, TValue> {
     return null;
   };
 
+  private isString = () => {
+    this.pushRule(new IsString());
+    return new StringRuleValidator<TModel, TValue>(this.pushRule)
```
[the remark this answers is not in the document]
Comment (AndreaCimini90) [reply packages/pn-validator/src/ValidatorBuilder.ts:47]: Added

File: packages/pn-validator/src/__test__/ValidatorBuilder.test.ts [hunk 6]
```diff
@@ -17,65 +17,70 @@ describe('Test ValidatorBuilder', () => {
   it('check if getTypeRules returns correct rules (string)', () => {
     const dummyValidatorBuilder = new ValidatorBuilder<any, String>();
     const rules = dummyValidatorBuilder.getTypeRules();
-    expect(rules.isEmpty).toBeDefined();
-    expect(rules.length).toBeDefined();
-    expect(rules.matches).toBeDefined();
-    checkCommonRules(rules);
+    expect(rules.isString).toBeDefined();
```
[the remark this answers is not in the document]
Comment (AndreaCimini90) [reply packages/pn-validator/src/__test__/ValidatorBuilder.test.ts:20]: done

File: packages/pn-validator/src/rules/ForEachElement.ts (hunk 1 above)
Comment (AndreaCimini90) [reply packages/pn-validator/src/rules/ForEachElement.ts:97]: yes, you are right!

File: packages/pn-validator/src/types/TypeRules.ts (hunk 2 above)
Comment (AndreaCimini90) [reply packages/pn-validator/src/types/TypeRules.ts:12]: Done

**Review by clombardi — APPROVED** [review]
OK for me, ready to merge.

**Pull request merged**